contracts/zaps/ZapSteth.sol

```diff
@@ -0,0 +1,309 @@
+    function zapOutEth(
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _slippageAllowance,
+        int128 _zeroIfPrincipal
+    ) external {
+        _zapOut(
+            _amount,
+            _expiration,
+            _position,
+            _slippageAllowance,
+            _zeroIfPrincipal,
+            0
+        );
+    }
+
+    /// @notice This function takes principal or yield tokens and redeems them for stETH.
+    /// @param _amount Amount of PT or YT to withdraw.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _slippageAllowance The slippage allowed for the internal curve stable-swap call.
+    /// @param _zeroIfPrincipal Zero if the assets to redeem are principal tokens.
+    function zapOutStEth(
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _slippageAllowance,
+        int128 _zeroIfPrincipal
+    ) external {
+        _zapOut(
+            _amount,
+            _expiration,
+            _position,
+            _slippageAllowance,
+            _zeroIfPrincipal,
+            1
+        );
+    }
+
+    /// @notice There should never be any tokens in this contract.
+    /// This function can rescue any possible leftovers.
+    /// @param token The token to rescue.
+    /// @param amount The amount to rescue.
+    function rescueTokens(address token, uint256 amount) external onlyOwner {
+        if (token == address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)) {
+            amount = Math.min(address(this).balance, amount);
+            payable(msg.sender).transfer(amount);
+        } else {
+            IERC20 want = IERC20(token);
+            amount = Math.min(want.balanceOf(address(this)), amount);
+            want.transfer(msg.sender, amount);
+        }
+    }
+
+    /// @notice Helper function that takes principal or yield tokens and redeems them for ETH or stETH.
+    /// @param _amount Amount of PT or YT to withdraw.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _slippageAllowance The slippage allowed for the internal curve stable-swap call.
+    /// @param _zeroIfPrincipal Zero if the assets to redeem are principal tokens.
+    /// @param _zeroIfEth Zero if the assets to redeem for is ETH. otherwise it will be stETH.
+    function _zapOut(
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _slippageAllowance,
+        int128 _zeroIfPrincipal,
+        int128 _zeroIfEth
+    ) internal {
+        ITranche tranche = _deriveTranche(address(_position), _expiration);
+
+        if (_zeroIfPrincipal == 0) {
+            tranche.transferFrom(msg.sender, address(this), _amount);
+            tranche.approve(address(tranche), _amount);
```

Comment:
I wonder if there's a way we could factor this out into another method which sets max allowance on any tranche which can be derived via create2? The pro is that we don't have to approve on each call, the con is that this will need a transaction before it can work.

Reply:
i removed this entirely